File: packages/tasq-evals/fixtures/discovery-recipe-client.py
#!/usr/bin/env python3
"""Package-independent client: executes only argv recipes returned by its pointer."""

import json
import os
import subprocess
import sys


def run(argv):
    completed = subprocess.run(argv, env=os.environ.copy(), capture_output=True, text=True)
    try:
        payload = json.loads(completed.stdout)
    except json.JSONDecodeError as error:
        raise RuntimeError(f"command returned non-JSON stdout: {completed.stdout!r}; stderr={completed.stderr!r}") from error
    return {"exitCode": completed.returncode, "stdout": payload, "stderr": completed.stderr}


request = json.load(sys.stdin)
bootstrap_run = run(request["pointerArgv"])
if bootstrap_run["exitCode"] != 0:
    print(json.dumps({"bootstrap": bootstrap_run, "results": []}, ensure_ascii=False))
    sys.exit(0)

bootstrap = bootstrap_run["stdout"]
if bootstrap.get("contractVersion") != "tasq.autonomous-bootstrap.v1" or not isinstance(bootstrap.get("recipes"), list):
    raise RuntimeError(f"unsupported or malformed bootstrap contract: {bootstrap.get('contractVersion')!r}")
results = []
for action in request.get("actions", []):
    selector = action.get("selector")
    if selector is None:
        matches = [item for item in bootstrap["recipes"] if item.get("id") == action["recipeId"]]
        selection_label = action["recipeId"]
    else:
        parameter_names = sorted(selector.get("parameterNames", []))
        matches = [item for item in bootstrap["recipes"] if
                   item.get("outputContract") == selector.get("outputContract") and
                   item.get("mutates") == selector.get("mutates") and
                   (selector.get("requiredCapability") is None or
                    item.get("requiredCapability") == selector.get("requiredCapability")) and
                   sorted(parameter.get("name") for parameter in item.get("parameters", [])) == parameter_names]
        selection_label = json.dumps(selector, sort_keys=True)
    if len(matches) != 1:
        raise RuntimeError(f"discovery must advertise exactly one recipe matching {selection_label}")
    recipe = matches[0]
    if recipe.get("version") != 1 or not isinstance(recipe.get("argvTemplate"), list) or not isinstance(recipe.get("parameters"), list):
        raise RuntimeError(f"unsupported or malformed selected recipe {recipe.get('id')!r}")
    replacements = action.get("replacements", {})
    declared = {parameter["placeholder"] for parameter in recipe["parameters"]}
    if declared != set(replacements):
        raise RuntimeError(f"recipe parameters differ: declared={declared}, supplied={set(replacements)}")
    argv = [replacements.get(part, part) for part in recipe["argvTemplate"]]
    results.append({"selectedRecipeId": recipe["id"], **run(argv)})

print(json.dumps({"bootstrap": bootstrap_run, "results": results}, ensure_ascii=False))

File: packages/tasq-evals/fixtures/discovery-recipe-client.mjs
#!/usr/bin/env node
/** Package-independent JS client: knows JSON/processes, not Tasq packages. */

import { spawnSync } from "node:child_process";

const chunks = [];
for await (const chunk of process.stdin) chunks.push(chunk);
const request = JSON.parse(Buffer.concat(chunks).toString("utf8"));

function run(argv) {
  // Maximum legal space/actor identifiers are repeated in each executable
  // recipe. Keep the independent client bounded, but above Node's
  // platform-dependent synchronous pipe buffer so valid bootstrap JSON is
  // never truncated before parsing.
  const child = spawnSync(argv[0], argv.slice(1), {
    env: process.env,
    encoding: "utf8",
    maxBuffer: 4 * 1024 * 1024,
  });
  let stdout;
  try {
    stdout = JSON.parse(child.stdout);
  } catch (error) {
    throw new Error(`command returned non-JSON stdout: ${JSON.stringify(child.stdout)}; stderr=${JSON.stringify(child.stderr)}`, { cause: error });
  }
  return { exitCode: child.status, stdout, stderr: child.stderr };
}

const bootstrap = run(request.pointerArgv);
const results = [];
if (bootstrap.exitCode === 0) {
  if (bootstrap.stdout?.contractVersion !== "tasq.autonomous-bootstrap.v1" || !Array.isArray(bootstrap.stdout.recipes)) {
    throw new Error(`unsupported or malformed bootstrap contract: ${String(bootstrap.stdout?.contractVersion)}`);
  }
  for (const action of request.actions ?? []) {
    const selector = action.selector;
    const matches = selector
      ? bootstrap.stdout.recipes.filter((candidate) =>
          candidate.outputContract === selector.outputContract &&
          candidate.mutates === selector.mutates &&
          (selector.requiredCapability === undefined ||
            candidate.requiredCapability === selector.requiredCapability) &&
          JSON.stringify(candidate.parameters.map((parameter) => parameter.name).sort()) ===
            JSON.stringify([...(selector.parameterNames ?? [])].sort()))
      : bootstrap.stdout.recipes.filter((candidate) => candidate.id === action.recipeId);
    const selectionLabel = selector ? JSON.stringify(selector) : action.recipeId;
    if (matches.length !== 1) throw new Error(`discovery must advertise exactly one recipe matching ${selectionLabel}`);
    const recipe = matches[0];
    if (recipe.version !== 1 || !Array.isArray(recipe.argvTemplate) || !Array.isArray(recipe.parameters)) {
      throw new Error(`unsupported or malformed selected recipe ${String(recipe.id)}`);
    }
    const declared = recipe.parameters.map((parameter) => parameter.placeholder).sort();
    const supplied = Object.keys(action.replacements ?? {}).sort();
    if (JSON.stringify(declared) !== JSON.stringify(supplied)) {
      throw new Error(`recipe parameters differ: declared=${declared}, supplied=${supplied}`);
    }
    const argv = recipe.argvTemplate.map((part) => action.replacements[part] ?? part);
    results.push({ selectedRecipeId: recipe.id, ...run(argv) });
  }
}

process.stdout.write(JSON.stringify({ bootstrap, results }) + (request.holdAfterResult ? "\n" : ""));
if (request.holdAfterResult) setInterval(() => {}, 60_000);
